File: MercaderSG/Comercial/BuscarProducto.vb
Imports Negocios
Imports Servicios
Imports Entidades
Imports Excepciones
Imports System.IO

Public Class BuscarProducto

    Dim ListaProductos As New List(Of ProductoEN)
    Dim ListaProductosTemp As New List(Of ProductoEN)
    Dim ListaProductosTempGral As New List(Of ProductoEN)

    Dim NroPag As Integer = 0

    Private _productosel As ProductoEN
    Public Property ProductoSel() As ProductoEN
        Get
            Return _productosel
        End Get
        Set(ByVal value As ProductoEN)
            _productosel = value
        End Set
    End Property


    Private Sub BuscarProducto_Load(sender As System.Object, e As System.EventArgs) Handles MyBase.Load
        BGW.WorkerReportsProgress = True
        BGW.WorkerSupportsCancellation = True
        BGW.RunWorkerAsync()

        BarraProgreso.ShowDialog()

        CargarTT()
        AplicarIdioma()
        CargarFoco(ProductoGB)

        BuscarCmb.Items.Add(My.Resources.ArchivoIdioma.CMBNombre)
        BuscarCmb.Items.Add(My.Resources.ArchivoIdioma.CMBSector)

        BuscarCmb.SelectedIndex = 0

        ListaProductosTempGral.AddRange(ListaProductos)
    End Sub

    Private Sub BGW_DoWork(sender As System.Object, e As System.ComponentModel.DoWorkEventArgs) Handles BGW.DoWork
        ListaProductos = ProductoRN.CargarProducto()
        For i As Integer = 0 To ListaProductos.Count - 1
            If BGW.CancellationPending Then
                e.Cancel = True
                Exit For
            Else
                BGW.ReportProgress(100 * i / ListaProductos.Count)
            End If
        Next
    End Sub

    Private Sub BGW_ProgressChanged(sender As System.Object, e As System.ComponentModel.ProgressChangedEventArgs) Handles BGW.ProgressChanged
        BarraProgreso.ProgressBar1.Value = e.ProgressPercentage
    End Sub

    Private Sub BGW_RunWorkerCompleted(sender As System.Object, e As System.ComponentModel.RunWorkerCompletedEventArgs) Handles BGW.RunWorkerCompleted
        ProductosDG.AutoGenerateColumns = False
        ProductosDG.DataSource = PaginaSig(NroPag)

        BarraProgreso.ProgressBar1.Value = 0
        BarraProgreso.Close()
    End Sub

    Function PaginaSig(NroPagActual As Integer) As List(Of ProductoEN)
        NroPag += 1

        Dim ListaAux As New List(Of ProductoEN)
        For Each item As ProductoEN In ListaProductos.Paginacion(NroPag)
            Dim UnProducto As New ProductoEN

            UnProducto.CodProd = item.CodProd
            UnProducto.Nombre = item.Nombre
            UnProducto.Sector = item.Sector
            UnProducto.Cantidad = item.Cantidad
            UnProducto.Precio = item.Precio
            UnProducto.Descripcion = item.Descripcion

            ListaAux.Add(UnProducto)
        Next

        Return ListaAux
    End Function

    Private Sub ProductosDG_CellDoubleClick(sender As System.Object, e As System.Windows.Forms.DataGridViewCellEventArgs) Handles ProductosDG.CellDoubleClick
        _productosel = DirectCast(ProductosDG.CurrentRow.DataBoundItem, ProductoEN)
        DialogResult = Windows.Forms.DialogResult.OK
    End Sub

    Private Sub BuscarBtn_Click(sender As System.Object, e As System.EventArgs) Handles BuscarBtn.Click

        If Not ConsistenciaDatos() Then
            Exit Sub
        End If

        NroPag = 0

        ListaProductosTemp.Clear()
        ListaProductosTemp.AddRange(ListaProductos)

        ListaProductos.Clear()

        Try
            ListaProductos = ProductoRN.BuscarProducto(BuscarCmb.SelectedItem, BusquedaTxt.Text)
        Catch ex As WarningException
            MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)
            BusquedaTxt.Clear()

            ListaProductos.Clear()
            ListaProductos.AddRange(ListaProductosTempGral)

            NroPag = 0

            ProductosDG.DataSource = Nothing
            ProductosDG.DataSource = PaginaSig(NroPag)
        End Try

        ProductosDG.DataSource = Nothing
        ProductosDG.AutoGenerateColumns = False
        ProductosDG.DataSource = PaginaSig(NroPag)

        If ProductosDG.Rows.Count = 0 Then
            MessageBox.Show(My.Resources.ArchivoIdioma.NoExisteProdBusqueda, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning)
            BusquedaTxt.Clear()

            ListaProductos.Clear()
            ListaProductos.AddRange(ListaProductosTempGral)

            NroPag = 0

            ProductosDG.DataSource = Nothing
            ProductosDG.DataSource = PaginaSig(NroPag)
        End If
    End Sub

    Private Sub RecargarBtn_Click(sender As System.Object, e As System.EventArgs) Handles RecargarBtn.Click
        ListaProductos.Clear()
        ListaProductos.AddRange(ListaProductosTempGral)

        NroPag = 0
        ProductosDG.DataSource = Nothing
        ProductosDG.DataSource = PaginaSig(NroPag)
    End Sub

    Private Sub BuscarProducto_KeyDown(sender As System.Object, e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyDown
        If e.KeyCode = Keys.Enter Then
            _productosel = DirectCast(ProductosDG.CurrentRow.DataBoundItem, ProductoEN)
            DialogResult = Windows.Forms.DialogResult.OK
        End If
        If e.KeyCode = Keys.Escape Then
            Me.Close()
        End If
    End Sub

#Region "Validaciones"

    Private Function ConsistenciaDatos() As Boolean
        Dim Resultado As Boolean = True

        If String.IsNullOrEmpty(BusquedaTxt.Text) Then
            MensajeTT.Show(My.Resources.ArchivoIdioma.BusquedaVacia, BusquedaTxt)
            BusquedaTxt.Clear()
            BusquedaTxt.Focus()
            Resultado = False
        End If

        Select Case BuscarCmb.SelectedItem
            Case My.Resources.ArchivoIdioma.CMBNombre

                If BusquedaTxt.Text.Length > 20 Then

                    MensajeTT.Show(My.Resources.ArchivoIdioma.Contener20Carac, BusquedaTxt)
                    BusquedaTxt.Focus()
                    Resultado = False

                End If

            Case My.Resources.ArchivoIdioma.CMBSector

                If BusquedaTxt.Text.Length > 50 Then

                    MensajeTT.Show(My.Resources.ArchivoIdioma.Contener50Carac, BusquedaTxt)
                    BusquedaTxt.Focus()
                    Resultado = False

                End If

        End Select

        Return Resultado
    End Function

#End Region

#Region "Idioma y TT"

    Public Sub AplicarIdioma()
        Me.Text = My.Resources.ArchivoIdioma.BuscarProductoFrm

        ProductoGB.Text = My.Resources.ArchivoIdioma.SeleccionarProd
        BuscarPorLbl.Text = My.Resources.ArchivoIdioma.BuscarPorLbl
        IgualLbl.Text = My.Resources.ArchivoIdioma.IgualALbl
        BuscarBtn.Text = My.Resources.ArchivoIdioma.BuscarBtn

        CodProdCAB.HeaderText = My.Resources.ArchivoIdioma.CodigoCAB
        NombreCAB.HeaderText = My.Resources.ArchivoIdioma.CMBNombre
        PrecioUnitarioCAB.HeaderText = My.Resources.ArchivoIdioma.Precio
        CantidadCAB.HeaderText = My.Resources.ArchivoIdioma.Cantidad
        SectorCAB.HeaderText = My.Resources.ArchivoIdioma.CMBSector
    End Sub

    Private Sub CargarTT()
        ControlesTP.SetToolTip(BusquedaTxt, My.Resources.ArchivoIdioma.IngresarProdBuscar)
        ControlesTP.SetToolTip(BuscarBtn, My.Resources.ArchivoIdioma.TTBuscarProd)
        ControlesTP.SetToolTip(RecargarBtn, My.Resources.ArchivoIdioma.TTRecargarBtn)
        ControlesTP.SetToolTip(ProductosDG, My.Resources.ArchivoIdioma.TTListaProducto)
    End Sub

#End Region

#Region "Mejoras"
    Private Sub TieneFoco(ByVal sender As System.Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.WhiteSmoke
    End Sub

    Private Sub PierdeFoco(ByVal sender As Object, ByVal e As System.EventArgs)
        Dim MiTextBox As TextBox = DirectCast(sender, TextBox)
        MiTextBox.BackColor = Color.White
    End Sub

    Public Sub CargarFoco(Grupo As Windows.Forms.GroupBox)
        Dim Ctrl As Control
        For Each Ctrl In Grupo.Controls
            If (TypeOf (Ctrl) Is System.Windows.Forms.TextBox) Then
                Dim MiTextBox As TextBox = DirectCast(Ctrl, TextBox)
                If MiTextBox.ReadOnly = False Then
                    AddHandler MiTextBox.GotFocus, AddressOf TieneFoco
                    AddHandler MiTextBox.LostFocus, AddressOf PierdeFoco
                End If
            End If
        Next
    End Sub
#End Region

End Class

File: MercaderSG/Comercial/BuscarProducto.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Entidades;
using Excepciones;
using Microsoft.VisualBasic.CompilerServices;
using Negocios;

namespace MercaderSG
{
    public partial class BuscarProducto
    {
        public BuscarProducto()
        {
            InitializeComponent();
            _RecargarBtn.Name = "RecargarBtn";
            _ProductosDG.Name = "ProductosDG";
            _BuscarBtn.Name = "BuscarBtn";
        }

        private List<ProductoEN> ListaProductos = new List<ProductoEN>();
        private List<ProductoEN> ListaProductosTemp = new List<ProductoEN>();
        private List<ProductoEN> ListaProductosTempGral = new List<ProductoEN>();
        private int NroPag = 0;
        private ProductoEN _productosel;

        public ProductoEN ProductoSel
        {
            get
            {
                return _productosel;
            }

            set
            {
                _productosel = value;
            }
        }

        private void BuscarProducto_Load(object sender, EventArgs e)
        {
            BGW.WorkerReportsProgress = true;
            BGW.WorkerSupportsCancellation = true;
            BGW.RunWorkerAsync();
            My.MyProject.Forms.BarraProgreso.ShowDialog();
            CargarTT();
            AplicarIdioma();
            CargarFoco(ProductoGB);
            BuscarCmb.Items.Add(My.Resources.ArchivoIdioma.CMBNombre);
            BuscarCmb.Items.Add(My.Resources.ArchivoIdioma.CMBSector);
            BuscarCmb.SelectedIndex = 0;
            ListaProductosTempGral.AddRange(ListaProductos);
        }

        private void BGW_DoWork(object sender, System.ComponentModel.DoWorkEventArgs e)
        {
            ListaProductos = ProductoRN.CargarProducto();
            for (int i = 0, loopTo = ListaProductos.Count - 1; i <= loopTo; i++)
            {
                if (BGW.CancellationPending)
                {
                    e.Cancel = true;
                    break;
                }
                else
                {
                    BGW.ReportProgress((int)(100 * i / (double)ListaProductos.Count));
                }
            }
        }

        private void BGW_ProgressChanged(object sender, System.ComponentModel.ProgressChangedEventArgs e)
        {
            My.MyProject.Forms.BarraProgreso.ProgressBar1.Value = e.ProgressPercentage;
        }

        private void BGW_RunWorkerCompleted(object sender, System.ComponentModel.RunWorkerCompletedEventArgs e)
        {
            ProductosDG.AutoGenerateColumns = false;
            ProductosDG.DataSource = PaginaSig(NroPag);
            My.MyProject.Forms.BarraProgreso.ProgressBar1.Value = 0;
            My.MyProject.Forms.BarraProgreso.Close();
        }

        public List<ProductoEN> PaginaSig(int NroPagActual)
        {
            NroPag += 1;
            var ListaAux = new List<ProductoEN>();
            foreach (ProductoEN item in ListaProductos.Paginacion(NroPag))
            {
                var UnProducto = new ProductoEN();
                UnProducto.CodProd = item.CodProd;
                UnProducto.Nombre = item.Nombre;
                UnProducto.Sector = item.Sector;
                UnProducto.Cantidad = item.Cantidad;
                UnProducto.Precio = item.Precio;
                UnProducto.Descripcion = item.Descripcion;
                ListaAux.Add(UnProducto);
            }

            return ListaAux;
        }

        private void ProductosDG_CellDoubleClick(object sender, DataGridViewCellEventArgs e)
        {
            _productosel = (ProductoEN)ProductosDG.CurrentRow.DataBoundItem;
            DialogResult = DialogResult.OK;
        }

        private void BuscarBtn_Click(object sender, EventArgs e)
        {
            if (!ConsistenciaDatos())
            {
                return;
            }

            NroPag = 0;
            ListaProductosTemp.Clear();
            ListaProductosTemp.AddRange(ListaProductos);
            ListaProductos.Clear();
            try
            {
                ListaProductos = ProductoRN.BuscarProducto(Conversions.ToString(BuscarCmb.SelectedItem), BusquedaTxt.Text);
            }
            catch (WarningException ex)
            {
                MessageBox.Show(ex.Message, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                BusquedaTxt.Clear();
                ListaProductos.Clear();
                ListaProductos.AddRange(ListaProductosTempGral);
                NroPag = 0;
                ProductosDG.DataSource = null;
                ProductosDG.DataSource = PaginaSig(NroPag);
            }

            ProductosDG.DataSource = null;
            ProductosDG.AutoGenerateColumns = false;
            ProductosDG.DataSource = PaginaSig(NroPag);
            if (ProductosDG.Rows.Count == 0)
            {
                MessageBox.Show(My.Resources.ArchivoIdioma.NoExisteProdBusqueda, My.Resources.ArchivoIdioma.MsgBoxAdvertencia, MessageBoxButtons.OK, MessageBoxIcon.Warning);
                BusquedaTxt.Clear();
                ListaProductos.Clear();
                ListaProductos.AddRange(ListaProductosTempGral);
                NroPag = 0;
                ProductosDG.DataSource = null;
                ProductosDG.DataSource = PaginaSig(NroPag);
            }
        }

        private void RecargarBtn_Click(object sender, EventArgs e)
        {
            ListaProductos.Clear();
            ListaProductos.AddRange(ListaProductosTempGral);
            NroPag = 0;
            ProductosDG.DataSource = null;
            ProductosDG.DataSource = PaginaSig(NroPag);
        }

        private void BuscarProducto_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                _productosel = (ProductoEN)ProductosDG.CurrentRow.DataBoundItem;
                DialogResult = DialogResult.OK;
            }

            if (e.KeyCode == Keys.Escape)
            {
                Close();
            }
        }

        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private bool ConsistenciaDatos()
        {
            bool Resultado = true;
            if (string.IsNullOrEmpty(BusquedaTxt.Text))
            {
                MensajeTT.Show(My.Resources.ArchivoIdioma.BusquedaVacia, BusquedaTxt);
                BusquedaTxt.Clear();
                BusquedaTxt.Focus();
                Resultado = false;
            }

            switch (BuscarCmb.SelectedItem)
            {
                case var @case when Operators.ConditionalCompareObjectEqual(@case, My.Resources.ArchivoIdioma.CMBNombre, false):
                    {
                        if (BusquedaTxt.Text.Length > 20)
                        {
                            MensajeTT.Show(My.Resources.ArchivoIdioma.Contener20Carac, BusquedaTxt);
                            BusquedaTxt.Focus();
                            Resultado = false;
                        }

                        break;
                    }

                case var case1 when Operators.ConditionalCompareObjectEqual(case1, My.Resources.ArchivoIdioma.CMBSector, false):
                    {
                        if (BusquedaTxt.Text.Length > 50)
                        {
                            MensajeTT.Show(My.Resources.ArchivoIdioma.Contener50Carac, BusquedaTxt);
                            BusquedaTxt.Focus();
                            Resultado = false;
                        }

                        break;
                    }
            }

            return Resultado;
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public void AplicarIdioma()
        {
            Text = My.Resources.ArchivoIdioma.BuscarProductoFrm;
            ProductoGB.Text = My.Resources.ArchivoIdioma.SeleccionarProd;
            BuscarPorLbl.Text = My.Resources.ArchivoIdioma.BuscarPorLbl;
            IgualLbl.Text = My.Resources.ArchivoIdioma.IgualALbl;
            BuscarBtn.Text = My.Resources.ArchivoIdioma.BuscarBtn;
            CodProdCAB.HeaderText = My.Resources.ArchivoIdioma.CodigoCAB;
            NombreCAB.HeaderText = My.Resources.ArchivoIdioma.CMBNombre;
            PrecioUnitarioCAB.HeaderText = My.Resources.ArchivoIdioma.Precio;
            CantidadCAB.HeaderText = My.Resources.ArchivoIdioma.Cantidad;
            SectorCAB.HeaderText = My.Resources.ArchivoIdioma.CMBSector;
        }

        private void CargarTT()
        {
            ControlesTP.SetToolTip(BusquedaTxt, My.Resources.ArchivoIdioma.IngresarProdBuscar);
            ControlesTP.SetToolTip(BuscarBtn, My.Resources.ArchivoIdioma.TTBuscarProd);
            ControlesTP.SetToolTip(RecargarBtn, My.Resources.ArchivoIdioma.TTRecargarBtn);
            ControlesTP.SetToolTip(ProductosDG, My.Resources.ArchivoIdioma.TTListaProducto);
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private void TieneFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.WhiteSmoke;
        }

        private void PierdeFoco(object sender, EventArgs e)
        {
            TextBox MiTextBox = (TextBox)sender;
            MiTextBox.BackColor = Color.White;
        }

        public void CargarFoco(GroupBox Grupo)
        {
            foreach (Control Ctrl in Grupo.Controls)
            {
                if (Ctrl is TextBox)
                {
                    TextBox MiTextBox = (TextBox)Ctrl;
                    if (MiTextBox.ReadOnly == false)
                    {
                        MiTextBox.GotFocus += TieneFoco;
                        MiTextBox.LostFocus += PierdeFoco;
                    }
                }
            }
        }
        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }
}